### DISCUSSION TITLE
Simplify connection to sbt server?

### DISCUSSION THREAD

**Discussion by adpi2** [scalameta/metals#3255] [General]
When the user wants to connect Metals to sbt server it can use two commands which are slightly different:
- `BspSwitch`: This command only works if the `.bsp/sbt.json` file is already available. The user can then select `sbt` and Metals will connect to it. If the `.bsp/sbt.json` is absent, Metals will not propose to connect to sbt, which is source of confusion or disapointment.
- `GenerateBspConfig`: This command forces the generation of a `.bsp/sbt.json` file by running `sbt bspConfig`. Then Metals read the generated `.bsp/sbt.json` file to connect to sbt server. This method is slower than the first one because sbt is loaded twice.

This situation is not ideal because:
- the first command can fail (not propose sbt) and the second is generally slower.
- 2 commands is more complex than 1: we assume that the user knows how connection to sbt via BSP works (generation of the `.bsp/sbt.json` file then read and run the `argv` command`).

Ideally the user should not need to understand that complexity and it should be able to use the same command in every situation.

I propose to make `BspSwitch` the default command to connect to sbt server:
If there is no `.bsp/sbt.json` file, Metals will check the version of sbt (it must be bigger than 1.4.1) and if it is it will suggest to connect to sbt. After the user chooses `sbt` and because `.bsp/sbt.json` is absent, it will run `sbt bspConfig` and connect to sbt server.

`GenerateBspConfig` will still be useful to force the generation of a new BSP config, but it would rarely be used.

**Comment by dos65** [comment]
Sounds good!
I agree that Metals should provide an option to select sbt-bsp from a start even if bsp-config wasn't generated.

**Comment by ckipp01** [comment]
👍🏼 I think this would be good to get rid of the extra step.

**Comment by ckipp01** [comment]
Hey @adpi2 just letting you know I started on this. I have working pretty much what you explained.

- The `metals.generate-bsp-config` command can still be used the same to generate and then connect.
- The `metals.bsp-switch` command now works a bit different in that it doesn't just search the `.bsp/` but rather takes a look at whether the build tool can also be a build server and displays that option. You can see an example here:

[Video on Imgur](https://i.imgur.com/Q9LtIrG.mp4)

Notice there is no `.bsp/sbt.json` entry, but you still see sbt as an option when you trigger the bsp switch.

**Reply by ckipp01** [reply]
Just tying this all together, this was now merged in https://github.com/scalameta/metals/pull/3304. It works for both sbt and mill.

**Reply by adpi2** [reply]
Amazing thanks!
